include/boost/geometry/algorithms/simplify.hpp

```diff
@@ -116,34 +117,174 @@ struct simplify_range
+                max_distance = dist;
+                index = i;
+            }
+        }
+        return index;
+    }
+
+public :
+    template <typename Ring, typename Strategy, typename Distance>
+    static inline void apply(Ring const& ring, Ring& out,
+                    Distance const& max_distance, Strategy const& strategy)
+    {
+        std::size_t const size = boost::size(ring);
+        if (size == 0)
+        {
+            return;
+        }
+
+        int const input_sign = area_sign(geometry::area(ring));
+
+        std::set<std::size_t> visited_indexes;
+
+        // Rotate it into a copied vector
+        // (vector, because source type might not support rotation)
+        // (duplicate end point will be simplified away)
+        typedef typename geometry::point_type<Ring>::type point_type;
+
+        std::vector<point_type> rotated(size);
+
+        // Closing point (but it will not start here)
+        std::size_t index = 0;
+
+        // Iterate (usually one iteration is enough)
+        for (std::size_t iteration = 0; iteration < 4u; iteration++)
+        {
+            // Always take the opposite. Opposite guarantees that no point
+            // "halfway" is chosen, creating an artefact (very narrow triangle)
+            // Iteration 0: opposite to closing point (1/2, = on convex hull)
+            //              (this will start simplification with that point
+            //               and its opposite ~0)
+            // Iteration 1: move a quarter on that ring, then opposite to 1/4
+            //              (with its opposite 3/4)
+            // Iteration 2: move an eight on that ring, then opposite (1/8)
+            // Iteration 3: again move a quarter, then opposite (7/8)
+            // So finally 8 "sides" of the ring have been examined (if it were
+            // a semi-circle). Most probably, there are only 0 or 1 iterations.
+            switch (iteration)
+            {
+                case 1 : index = (index + size / 4) % size; break;
+                case 2 : index = (index + size / 8) % size; break;
+                case 3 : index = (index + size / 4) % size; break;
+            }
+            index = get_opposite<Strategy>(index, ring);
+
+            if (visited_indexes.count(index) > 0)
+            {
+                // Avoid trying the same starting point more than once
+                continue;
+            }
+
+            std::rotate_copy(boost::begin(ring), boost::begin(ring) + index,
+                             boost::end(ring), rotated.begin());
+
+            // Close the rotated copy
+            rotated.push_back(range::at(ring, index));
+
+            simplify_range<0>::apply(rotated, out, max_distance, strategy);
+
+            // Verify that what was positive, stays positive (or goes to 0)
+            // and what was negative stays negative (or goes to 0)
+            int const output_sign = area_sign(geometry::area(out));
+            if (output_sign == input_sign)
```

Comment:
Is this the check for the same orientation? If that's the case maybe we should have a function for that? To not use area explicitly because the actual area is not used here.

Reply:
Indeed, it checks for orientation only, that what was positive should stay positive, and what was negative should stay negative.

Reply:
I rather was thinking about making the code more self-descriptive. The area is not important here and may be confusing. The orientation is checked so if the function name reflected that then the coment wouldn't be needed because the code itself would make it clear. Furthermore enum type would be more appropriate than `int` to store the result. Therefore it could be something like `get_ring_orientation(a)` returning value of type `bg::order_selector` so `bg::clockwise`, `bg::counterclockwise ` or `bg::order_undetermined`.

Reply:
That is a good idea indeed.